x-pack/platform/plugins/shared/fleet/server/services/agentless/deployment_sync.ts

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Logger } from '@kbn/logging';
+import pMap from 'p-map';
+
+import type { AgentlessAgentService } from '../agents/agentless_agent';
+import { agentPolicyService, appContextService } from '..';
+
+import { AGENT_POLICY_SAVED_OBJECT_TYPE, SO_SEARCH_LIMIT } from '../../constants';
+
+const PAGE_SIZE = 20;
+
+function dryRunTag(dryRun = false) {
+  if (dryRun) {
+    return `[Dry Run]`;
+  }
+  return '';
+}
+
+export async function syncAgentlessDeployments(
+  {
+    logger,
+    agentlessAgentService,
+  }: {
+    logger: Logger;
+    agentlessAgentService: AgentlessAgentService;
+  },
+  opts?: {
+    dryRun?: boolean;
+    abortController?: AbortController;
+  }
+) {
+  logger.info(`[Agentless Deployment Sync] Starting sync process`);
+  const soClient = appContextService.getInternalUserSOClientWithoutSpaceExtension();
+
+  try {
+    // retrieve all agentless deployments
+    const currentDeployments: Array<{ policy_id: string; revision_idx?: number }> = [];
+    let hasMore = false;
+    let nextPageToken: string | undefined;
+
+    while (hasMore) {
+      const deploymentRes = await agentlessAgentService.listAgentlessDeployments({
+        perPage: PAGE_SIZE,
+        nextPageToken,
+      });
+
+      nextPageToken = deploymentRes.nextPageToken;
+      if (!nextPageToken) {
+        hasMore = false;
+      }
+
+      const policyIds = deploymentRes.deployments.map((d) => d.policy_id);
+      const agentPolicies = await agentPolicyService.getByIds(soClient, policyIds, {
+        spaceId: '*',
+        ignoreMissing: true,
+      });
+
+      await pMap(
+        deploymentRes.deployments,
+        async (deployment) => {
+          const agentPolicy = agentPolicies.find((ap) => ap.id === deployment.policy_id);
+          if (!agentPolicy) {
+            logger.info(
+              `[Agentless Deployment Sync]${dryRunTag(opts?.dryRun)} Deleting deployment ${
+                deployment.policy_id
+              }`
+            );
+            if (!opts?.dryRun) {
+              await agentlessAgentService
+                .deleteAgentlessAgent(deployment.policy_id)
+                .catch((error) => {
+                  logger.error(
+                    `[Agentless Deployment Sync] Failed to delete deployment ${deployment.policy_id}`,
+                    { error }
+                  );
+                });
+            }
+          } else if (
+            deployment.revision_idx === undefined ||
```

Comment:
so we update deployment just to have revision id?
i see this could happen in bulk for first run for every deployment, any issues that you see with that?

Reply:
> so we update deployment just to have revision id?
I see this could happen in bulk for first run for every deployment, any issues that you see with that?

Yes we update to have revision idx, on Kibana side, I do not think this will be an issue, what wil happens on agentless API/controller you will had a field or a label to the app config right? it will not update the actual deployment?